### PR TITLE
feat: Add comprehensive GRPO training system with dual GPU support and custom dataset integration

### PR DESCRIPTION
**Pull request by himanshushukla12**
# 🚀 GRPO Training System Enhancement

This PR introduces a complete GRPO (Group Relative Policy Optimization) training system with optimized GPU configurations, custom dataset support, and comprehensive documentation.

### 🎯 Core Features
- **Dual GPU Training Setup**: Optimized configuration for GPU 0 (vLLM) + GPU 1 (training)
- **Custom Dataset Integration**: Support for instruction-response format datasets
- **Memory Optimization**: Configurations for different GPU memory constraints
- **Ubuntu Compatibility**: Tested and optimized for Ubuntu environments

#### **Custom Dataset Transforms**
- `instruction_response_transform`: Basic JSONL processing for instruction-response pairs
- `instruction_response_with_tests_transform`: Enhanced validation with function/class detection
- Support for custom dataset formats with proper error handling

#### **Training Configurations**
- `custom_dataset_config.yaml`: Dual GPU setup (vLLM server + training process)
- `single_gpu_config.yaml`: Single GPU shared memory configuration
- `accelerate_config.yaml`: Single process setup to avoid distributed training issues

#### **Training Scripts**
- `train_single_gpu.sh`: Training process with GPU allocation and connectivity checks
- `start_vllm_server.sh`: vLLM server startup with proper GPU assignment
- `launch_training.py`: Python launcher with automated health checks

#### **Validation Tools**
- `simple_test_transforms.py`: Dataset transform validation script
- Environment setup and CUDA device management utilities

### **Comprehensive Guide (`how to.md`)**
- Complete setup instructions for single, dual, and multi-GPU configurations
- Custom dataset preparation with format specifications and examples
- Dataset transform creation and validation procedures
- Memory optimization strategies and troubleshooting guide
- Step-by-step terminal workflows for all GPU configurations

### **Enhanced README**
- Updated with dual GPU setup instructions
- Memory optimization and error resolution guides
- Cross-references with the comprehensive how-to guide

### **Memory Management**
- Optimized batch sizes and gradient accumulation for different GPU configurations
- CUDA memory allocation improvements with `expandable_segments:True`
- Proper GPU device assignment to prevent context conflicts

### **Error Resolution**
- Fixed DTensor mixed with Tensor errors through proper accelerate configuration
- Resolved CUDA device initialization issues
- Eliminated tokenizer parallelism warnings

### **Performance Optimization**
- Dedicated GPU allocation (GPU 0: vLLM, GPU 1: Training)
- Optimized vLLM server configuration with aggressive memory utilization
- Efficient data loading with reduced memory footprint

## 🧪 Testing & Validation

- ✅ Successfully tested on Ubuntu with dual RTX 6000 Ada Generation GPUs
- ✅ Validated custom dataset transforms with sample data
- ✅ Confirmed vLLM server connectivity and training process integration
- ✅ Memory optimization prevents CUDA out-of-memory errors

## 📊 Results

The training system now successfully:
- Loads custom instruction-response datasets (68K+ samples tested)
- Runs stable GRPO training with proper reward calculation
- Utilizes both format and execution reward functions
- Maintains separate GPU contexts for optimal performance

**Sample Training Output:**
Terminal 1 (training):
```bash
{'loss': 0.0001, 'grad_norm': 0.0098876953125, 'learning_rate': 5.3e-06, 'num_tokens': 5514904.0, 'completion_length': 110.40625, 'rewards/soft_format_reward_func': 0.25, 'rewards/code_execution_reward_func': 0.5, 'reward': 0.75, 'reward_std': 0.0, 'kl': 0.12127423752099276, 'clip_ratio': 0.0, 'epoch': 0.05}
{'loss': 0.0002, 'grad_norm': 0.015869140625, 'learning_rate': 5.3e-06, 'num_tokens': 5527158.0, 'completion_length': 110.9375, 'rewards/soft_format_reward_func': 0.25, 'rewards/code_execution_reward_func': 0.5, 'reward': 0.75, 'reward_std': 0.0, 'kl': 0.15593589656054974, 'clip_ratio': 0.0, 'epoch': 0.05}
{'loss': 0.0002, 'grad_norm': 0.058349609375, 'learning_rate': 5.3e-06, 'num_tokens': 5539474.0, 'completion_length': 112.0, 'rewards/soft_format_reward_func': 0.25, 'rewards/code_execution_reward_func': 0.5, 'reward': 0.75, 'reward_std': 0.0, 'kl': 0.2036700863391161, 'clip_ratio': 0.0, 'epoch': 0.05}
```

Terminal 2 (vLLM):

```bash
NFO:     127.0.0.1:56414 - "POST /update_named_param/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:56414 - "POST /reset_prefix_cache/ HTTP/1.1" 200 OK
Processed prompts:  25%|██████████████████████████████████████                                                                                                                  | 1/4 [00:01<00:03,  1.13s/it, est. speed input: 235.40 toks/s, output: 317.40 toks/s]
INFO:     127.0.0.1:56414 - "POST /generate/ HTTP/1.1" 200 OK
Processed prompts:  25%|██████████████████████████████████████                                                                                                                  | 1/4 [00:01<00:05,  1.97s/it, est. speed input: 137.40 toks/s, output: 251.98 toks/s]
INFO:     127.0.0.1:56414 - "POST /generate/ HTTP/1.1" 200 OK
Processed prompts:  25%|██████████████████████████████████████                                                                                                                  | 1/4 [00:01<00:03,  1.09s/it, est. speed input: 243.69 toks/s, output: 330.72 toks/s]
INFO:     127.0.0.1:56414 - "POST /generate/ HTTP/1.1" 200 OK
Processed prompts:   0%|                                                                                                                                                                    | 0/4 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]
```

### CODE REVIEW

**Comment by himanshushukla12** [comment]
hello maintainers, waiting for your response for this PR

**Comment by salmanmohammadi** [comment]
Hi @himanshushukla12, thank you for taking the time to open this PR.

Unfortunately I don't think I'll be able to land this PR. This repository exists as a minimal and clean reference implementation for our associated blogpost.

 If you'd like to share this with the world please feel free to fork this repo and publish your modifications in your own fork.
